File: Sources/DocShot/Models/RecordingOptions.swift
import Foundation

/// Which audio sources a recording captures.
///
/// R1 is video-only and constructs `.none` exclusively; the remaining cases exist so the R3 audio
/// milestone extends this type rather than replacing it. `ScreenCaptureKitRecordingSession`
/// rejects any non-`.none` value rather than silently ignoring it.
public enum RecordingAudioMode: Equatable, Sendable {
    case none
    case system
    case microphone(deviceID: String?)
    case systemAndMicrophone(deviceID: String?)
}

/// How a recording is configured, decided before the stream starts.
public struct RecordingOptions: Equatable, Sendable {
    public var audio: RecordingAudioMode
    public var showsCursor: Bool
    /// Frames per second the stream may not exceed. R1 fixes this at 30 with no user control.
    public var maximumFrameRate: Int

    public static let defaultMaximumFrameRate = 30

    public init(audio: RecordingAudioMode, showsCursor: Bool, maximumFrameRate: Int) {
        self.audio = audio
        self.showsCursor = showsCursor
        self.maximumFrameRate = maximumFrameRate
    }

    /// The only configuration R1 produces.
    public static func videoOnly(
        showsCursor: Bool,
        maximumFrameRate: Int = RecordingOptions.defaultMaximumFrameRate
    ) -> RecordingOptions {
        RecordingOptions(audio: .none, showsCursor: showsCursor, maximumFrameRate: maximumFrameRate)
    }

    public var isVideoOnly: Bool {
        audio == .none
    }

    /// R3's first audio slice supports ScreenCaptureKit system audio only. Microphone and mixed
    /// recording require their own capability and timestamp-validation work.
    public var supportsCurrentAudioPipeline: Bool {
        audio == .none || audio == .system
    }

    /// Whether the stream captures system audio at all. With audio off, no audio is requested
    /// from ScreenCaptureKit, no audio writer input is created, and the file has no audio track.
    public var capturesSystemAudio: Bool {
        audio == .system
    }

    /// Whether DocShot's own process audio must be kept out of the system-audio track.
    ///
    /// The video filter excludes DocShot's windows, but that says nothing about audio: system
    /// audio is captured globally, for a window target as much as a display one. Without this,
    /// DocShot's own alert sounds and error beeps would be recorded into the user's clip.
    public var excludesOwnProcessAudio: Bool {
        capturesSystemAudio
    }
}

/// The deliberately fixed R4 GIF profile. GIF is a post-stop conversion of a completed MP4;
/// these limits are checked before presenting any user-facing Save panel.
public struct GIFProfile: Equatable, Sendable {
    public let maximumDuration: TimeInterval
    public let framesPerSecond: Int
    public let maximumLongEdge: Int
    public let maximumEncodedBytes: Int

    public init(
        maximumDuration: TimeInterval,
        framesPerSecond: Int,
        maximumLongEdge: Int,
        maximumEncodedBytes: Int
    ) {
        self.maximumDuration = maximumDuration
        self.framesPerSecond = framesPerSecond
        self.maximumLongEdge = maximumLongEdge
        self.maximumEncodedBytes = maximumEncodedBytes
    }

    /// Owner-approved R4 export policy: 15 seconds, 10 fps, 960 px, 10 MB.
    public static let docShotR4 = GIFProfile(
        maximumDuration: 15,
        framesPerSecond: 10,
        maximumLongEdge: 960,
        maximumEncodedBytes: 10 * 1_000_000
    )

    public func supports(_ recording: TemporaryRecording) -> Bool {
        recording.isPlayable && recording.duration <= maximumDuration
    }
}

File: windows/src/DocShot.Core/Models/RecordingOptions.cs
namespace DocShot.Core.Models;

/// <summary>
/// Which audio sources a recording captures.
/// </summary>
/// <remarks>
/// On Windows there is no single API that captures video and audio on one stream/one clock the
/// way ScreenCaptureKit does. <see cref="System"/> and <see cref="Microphone"/> both require a
/// parallel WASAPI capture client muxed into the same Media Foundation sink writer - see the
/// platform mapping table and risk spike W0-3 in <c>docs/WINDOWS_PORT_PLAN.md</c>. This type
/// itself does not change shape for that; the difference lives entirely in
/// <c>DocShot.Platform</c>'s implementation.
/// </remarks>
public abstract record RecordingAudioMode
{
    public sealed record None : RecordingAudioMode;
    public sealed record System : RecordingAudioMode;
    public sealed record Microphone(string? DeviceId) : RecordingAudioMode;
    public sealed record SystemAndMicrophone(string? DeviceId) : RecordingAudioMode;

    private RecordingAudioMode() { }

    public static readonly RecordingAudioMode NoAudio = new None();
}

/// <summary>How a recording is configured, decided before the stream starts.</summary>
public sealed record RecordingOptions(
    RecordingAudioMode Audio,
    bool ShowsCursor,
    int MaximumFrameRate)
{
    public const int DefaultMaximumFrameRate = 30;

    /// <summary>The only configuration the video-only milestone produces.</summary>
    public static RecordingOptions VideoOnly(bool showsCursor, int maximumFrameRate = DefaultMaximumFrameRate) =>
        new(RecordingAudioMode.NoAudio, showsCursor, maximumFrameRate);

    public bool IsVideoOnly => Audio is RecordingAudioMode.None;

    /// <summary>Whether the stream captures system audio at all.</summary>
    public bool CapturesSystemAudio => Audio is RecordingAudioMode.System or RecordingAudioMode.SystemAndMicrophone;

    /// <summary>
    /// Whether DocShot's own process audio must be kept out of the system-audio track. System
    /// audio is captured globally, for a window target as much as a display one, so without this
    /// DocShot's own alert sounds would be recorded into the user's clip.
    /// </summary>
    public bool ExcludesOwnProcessAudio => CapturesSystemAudio;
}

/// <summary>
/// The GIF export profile. GIF is a post-stop conversion of a completed MP4; these limits are
/// checked before presenting any user-facing Save panel.
/// </summary>
public sealed record GifProfile(
    double MaximumDuration,
    int FramesPerSecond,
    int MaximumLongEdge,
    long MaximumEncodedBytes)
{
    /// <summary>Owner-approved profile carried over from macOS R4: 15 seconds, 10 fps, 960 px, 10 MB.</summary>
    public static readonly GifProfile DocShotDefault = new(
        MaximumDuration: 15,
        FramesPerSecond: 10,
        MaximumLongEdge: 960,
        MaximumEncodedBytes: 10_000_000);

    public bool Supports(TemporaryRecording recording) =>
        recording.IsPlayable && recording.Duration <= MaximumDuration;
}
